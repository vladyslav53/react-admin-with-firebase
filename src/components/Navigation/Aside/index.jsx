import React, { useState } from 'react';
import { useSelector, shallowEqual } from 'react-redux';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import NavLink from '../Link';
import paths from '../../../pages/Router/paths';
import classes from './Aside.module.scss';

export const SubMenu = ({ label, children }) => {
  const [active, setActive] = useState(false);

  return (
    <li className={classNames({ 'is-active': active })}>
      <a
        exact-active-class="is-active"
        className="has-icon has-dropdown-icon"
        onClick={() => setActive(!active)}
      >
        <span className="icon">
          <i className="mdi mdi-view-list" />
        </span>
        <span className="menu-item-label">{label}</span>
        <div className="dropdown-icon">
          <span className="icon">
            <i
              className={classNames(
                'mdi',
                { 'mdi-minus': active },
                { 'mdi-plus': !active }
              )}
            />
          </span>
        </div>
      </a>
      <ul>{children}</ul>
    </li>
  );
};

SubMenu.propTypes = {
  children: PropTypes.node.isRequired,
  label: PropTypes.string.isRequired
};

const Aside = ({ handleMobileToggle }) => {
  const { isAdmin } = useSelector(
    state => ({
      isAdmin: state.auth.userData.isAdmin
    }),
    shallowEqual
  );

  return (
    <aside className="aside is-placed-left is-expanded">
      <Link to={paths.ROOT} className="aside-tools">
        <div className="aside-tools-label">
          <span>
            <b>React</b> Firebase
          </span>
        </div>
      </Link>
      <div className="menu is-menu-main">
        <ul className="menu-list">
          <li>
            <NavLink
              to={paths.ROOT}
              className="has-icon"
              onClick={handleMobileToggle}
            >
              <span className="icon">
                <i className="mdi mdi-home" />
              </span>
              <span className="menu-item-label">Home</span>
            </NavLink>
          </li>
          {isAdmin && (
            <li>
              <NavLink
                to={paths.USERS}
                className="has-icon"
                onClick={handleMobileToggle}
              >
                <span className="icon">
                  <i className="mdi mdi-account-supervisor" />
                </span>
                <span className="menu-item-label">Users</span>
              </NavLink>
            </li>
          )}
          <SubMenu label="Dropdown Menu">
            <li>
              <NavLink
                className={classes.submenuLink}
                to={paths.SUBMENU_1}
                onClick={handleMobileToggle}
              >
                <span>Submenu 1</span>
              </NavLink>
            </li>
            <li>
              <NavLink
                className={classes.submenuLink}
                to={paths.SUBMENU_2}
                onClick={handleMobileToggle}
              >
                <span>Submenu 2</span>
              </NavLink>
            </li>
          </SubMenu>
        </ul>
      </div>
    </aside>
  );
};

Aside.propTypes = {
  handleMobileToggle: PropTypes.func.isRequired
};

export default Aside;
