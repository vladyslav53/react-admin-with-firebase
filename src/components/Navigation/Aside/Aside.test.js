import React from 'react';

import Aside, { SubMenu } from '.';
import NavLink from '../Link';
import paths from '../../../pages/Router/paths';

describe('<Aside /> rendering', () => {
  const onHandler = jest.fn();

  it('should render without crashing', () => {
    const component = shallowWithProvider(
      <Aside handleMobileToggle={onHandler} />
    )({
      auth: {
        userData: {
          isAdmin: true
        }
      }
    });

    expect(component).toMatchSnapshot();
  });

  it('should set the handleMobileToggle prop correctly', () => {
    const { component } = mountWithProvider(
      <Aside handleMobileToggle={onHandler} />
    )({
      auth: {
        userData: {
          isAdmin: true
        }
      }
    });

    expect(
      component
        .find(NavLink)
        .at(1)
        .prop('onClick')
    ).toEqual(onHandler);

    expect(
      component
        .find(NavLink)
        .at(0)
        .prop('onClick')
    ).toEqual(onHandler);

    expect(
      component
        .find(NavLink)
        .at(2)
        .prop('onClick')
    ).toEqual(onHandler);
  });

  it('should not render the /users link if it the user is not an admin', () => {
    const { component } = mountWithProvider(
      <Aside handleMobileToggle={onHandler} />
    )({
      auth: {
        userData: {
          isAdmin: false
        }
      }
    });

    expect(
      component.contains(
        <NavLink to={paths.USERS} onClick={onHandler}>
          Users
        </NavLink>
      )
    ).toBeFalsy();
  });

  it('should render the /users link if it the user is an admin', () => {
    const { component } = mountWithProvider(
      <Aside handleMobileToggle={onHandler} />
    )({
      auth: {
        userData: {
          isAdmin: true
        }
      }
    });

    expect(
      component.contains(
        <NavLink to={paths.USERS} onClick={onHandler}>
          Users
        </NavLink>
      )
    ).toBeFalsy();
  });

  it('should render the <SubMenu /> component if the user is an admin', () => {
    const { component } = mountWithProvider(
      <Aside handleMobileToggle={onHandler} />
    )({
      auth: {
        userData: {
          isAdmin: true
        }
      }
    });

    expect(component.exists(SubMenu)).toBeTruthy();
  });

  it('should not render the <SubMenu /> component if the user is not an admin', () => {
    const { component } = mountWithProvider(
      <Aside handleMobileToggle={onHandler} />
    )({
      auth: {
        userData: {
          isAdmin: false
        }
      }
    });

    expect(component.exists(SubMenu)).toBeFalsy();
  });
});
