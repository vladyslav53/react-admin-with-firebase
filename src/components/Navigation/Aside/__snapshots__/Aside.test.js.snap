// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Aside /> rendering should render without crashing 1`] = `
Object {
  "component": <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Aside
        handleMobileToggle={[MockFunction]}
      />
    </Provider>
  </Router>,
  "store": Object {
    "clearActions": [Function],
    "dispatch": [Function],
    "getActions": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
}
`;
